experimental/ast/decl.go

```diff
@@ -1,121 +1,235 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package ast
 
 import (
-	"fmt"
-
 	"github.com/bufbuild/protocompile/internal/arena"
 )
 
 const (
-	declEmpty declKind = iota + 1
-	declPragma
-	declPackage
-	declImport
-	declDef
-	declBody
-	declRange
+	DeclKindEmpty DeclKind = iota + 1
+	DeclKindSyntax
+	DeclKindPackage
+	DeclKindImport
+	DeclKindDef
+	DeclKindBody
+	DeclKindRange
 )
 
-// Decl is a Protobuf declaration.
+// DeclKind is a kind of declaration. There is one value of DeclKind for each
+// Decl* type in this package.
+type DeclKind int8
+
+// DeclAny is any Decl* type in this package.
+//
+// Values of this type can be obtained by calling an AsAny method on a Decl*
+// type, such as [DeclSyntax.AsAny]. It can be type-asserted back to any of
+// the concrete Decl* types using its own As* methods.
 //
-// This is implemented by types in this package of the form Decl*.
-type Decl interface {
-	Spanner
-
-	// with should be called on a nil value of this type (not
-	// a nil interface) and return the corresponding value of this type
-	// extracted from the given context and index.
-	//
-	// Not to be called directly; see rawDecl[T].With().
-	with(ctx *Context, ptr arena.Untyped) Decl
-
-	// kind returns what kind of decl this is.
-	declKind() declKind
-	// declIndex returns the untyped arena pointer for this declaration.
-	declIndex() arena.Untyped
+// This type is used in lieu of a putative Decl interface type to avoid heap
+// allocations in functions that would return one of many different Decl*
+// types.
+type DeclAny struct {
+	// NOTE: These fields are sorted by alignment.
```

Comment:
FWIW, I think there may be something we could add to the golangci config that might enforce that structs are laid out to minimize internal padding.